scalding-core/src/main/scala/com/twitter/scalding/Mode.scala

```diff
@@ -63,72 +75,258 @@ object Mode {
    * of a map/reduce job, open a given tap and return the TupleEntryIterator
    */
   def openForRead(config: Config, tap: Tap[_, _, _]): TupleEntryIterator
 
+  /** Create a new FlowConnector for this cascading planner */
+  def newFlowConnector(props: Config): FlowConnector
+
+  /**
+   * Create a new Flow using the default FlowConnector, and ensure any fabric-specific flow configuration is complete.
+   */
+  def newFlow(config: Config, flowDef: FlowDef): Flow[_] = {
+    val conf2 = setupCounterCreation(config)
+
+    val flowcon = newFlowConnector(conf2)
+    val flow = flowcon.connect(flowDef)
+    flow
+  }
+
+  /**
+   * @return true if a checkpoint is required before a hash join on the given platform
+   */
+  private[scalding] def getHashJoinAutoForceRight: Boolean = false
+
+  private[scalding] def defaultConfig: Config = Config.empty
+
+  /**
+   * @return a Config which includes the class name to use for creating Statistics Counters.
+   *         Please see the comment in CounterImpl for details (this is really fabric implementation details)
+   */
+  private[scalding] def setupCounterCreation(fp: Config): Config =
+    fp + (CounterImpl.CounterImplClass -> classOf[GenericFlowPCounterImpl].getCanonicalName)
+
+}
+
+trait StorageMode extends java.io.Serializable {
+  import StorageMode._
+
+  /**
+   * @return true if the file exists on the current filesystem.
+   */
+  def fileExists(filename: String): Boolean
+
+  def temporaryTypedSource[T]: TemporarySource[T]
+
+  /**
+   * Create a Tap for the purpose set out in {{readOrWrite}}, using the scheme provided by {{schemedSource}}
+   */
+  def createTap(schemedSource: SchemedSource, readOrWrite: AccessMode, mode: Mode, sinkMode: SinkMode): Tap[_, _, _]
+
+  /**
+   * Verify whether sources are valid, given the applicable source strictness mode
+   *
+   * @throws InvalidSourceException if sources are missing
+   */
+  def validateTap(schemedSource: SchemedSource): Unit
+
+  /**
+   * Create a memory-only tap for the purpose of spooling some tuples from memory.
+   * On Local modes, it is also possible to write to a memory tap.
+   */
+  def createMemoryTap(readOrWrite: AccessMode, fields: Fields, asBuffer: mutable.Buffer[Tuple]): Tap[_, _, _]
+
+  /** Create a write-only tap which simply discards the results */
+  def createNullTap: Tap[_, _, _]
+
+  // access this through the DistributedCacheFile object
+  private[scalding] def addCachedFile(file: UncachedFile): CachedFile
+
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def readFromFile(filename: String): String
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def writeToFile(filename: String, text: String): Unit
+}
+
+object StorageMode {
+  trait TemporarySource[T] {
+    def sink(conf: Config): TypedSink[T]
+    def downstreamPipe(conf: Config): TypedPipe[T]
+  }
+}
+
+trait Mode extends java.io.Serializable {
```

Comment:
can we add a comment about the requirement of constructors for reflective construction?